10_binary_counter.md

```diff
@@ -253,112 +251,121 @@ we just rebalanced parenthesis[^min-not-commutative].
 
         f(a, b) = f(b, a)
 
-    Informally, `f` does the same thing, regardless of the order of the inputs.
+    Informally, `f` gets the same result, regardless of the order of the inputs.
     For example, multiplication of integers is commutative:
 
         a * b = b * a
 
-    This fact about integers can be proven in the following manner:
-    If you arrange items into a block of columns and rows, such as: 
-
-        * * * * *
-        * * * * *
-        * * * * *
+    In Chapter 9.1 of "From Mathematics to Generic Programming", Alex gives
+    a neat visual proof of the commutativity of positive integers:
 
-    The number of elements does not change regardless of how it is rotated:
+                      * * *
+        * * * * *     * * *
+        * * * * *  =  * * *
+        * * * * *     * * *
+                      * * *
 
-        * * *
-        * * *
-        * * *
-        * * *
-        * * *
+    Or as [Dirchlet][dirichlet] put it: "Whether you arrange soldiers in rows or columns, you still have the same number of soldiers".
 
-    This proof is from [Dirichlet](https://en.wikipedia.org/wiki/Peter_Gustav_Lejeune_Dirichlet)
-    (see Chapter 9.1 of "From Mathematics to Generic Programming").
     An example of an operation which is not commutative is string concatenation. 
 
+        "Hello, " + "World!" != "World!" + "Hello, "
+
+[dirichlet]: https://en.wikipedia.org/wiki/Peter_Gustav_Lejeune_Dirichlet)
+
 ## Binary counting and reduction
 
 Here we come to the amazing idea of how to do this transformation.
-This is one of the most beautiful ideas which they
-kept secret from you.
+This is one of the most beautiful ideas which they kept secret from you.
 They should have taught it in high school.
-But, they want to publish papers
-themselves and not tell you the general mechanism.
+But, they want to publish papers themselves and not tell you the general mechanism.
+
+Let us assume we have elements of type `T` that need to be paired or combined in some way,
+whether with `min`, `+`, `merge`, or any other assocative operation on `T`.
+What we can do is create an array called a "counter".
```

Comment:
This information now reinforces that we are talking about generic element `T` with generic operation.